### PR TITLE
[ogr] Fix absence of proper feature ID when adding features to CSV, ODS, and XLSX datasets

### PR DESCRIPTION
**Pull request by nirvn**
## Description

This PR fixes the absence of a proper/valid feature ID when adding new features to preexisting [CSV,ODS,XLSX] datasets. 

@rouault , @nyalldawson , @suricactus , as discussed here (https://github.com/OSGeo/gdal/issues/3552).

### CODE REVIEW

**Comment by nirvn** [comment]
Unrelated test failure (https://cdash.orfeo-toolbox.org/index.php?project=QGIS&filtercount=1&showfilters=1&field1=buildname/string&compare1=63&value1=PR42189_7f0e1141c27eabea5f79d9498ccf230918a64e7d&date=2021-03-11):

```
The following tests FAILED:
	246 - qgis_3drenderingtest (SEGFAULT)
```

File: src/core/providers/ogr/qgsogrprovider.cpp [hunk 1]
```diff
@@ -1669,7 +1669,7 @@ QString QgsOgrProvider::jsonStringValue( const QVariant &value ) const
   return stringValue;
 }
 
-bool QgsOgrProvider::addFeaturePrivate( QgsFeature &f, Flags flags )
+bool QgsOgrProvider::addFeaturePrivate( QgsFeature &f, Flags flags, QgsFeatureId featureId )
```
Comment (rouault) [inline src/core/providers/ogr/qgsogrprovider.cpp:1672]:
```suggestion
// incrementalFeatureId will be generally -1, except for a few drivers where QGIS will set it to a likely value whereas OGR doesn't set it currently
bool QgsOgrProvider::addFeaturePrivate( QgsFeature &f, Flags flags, QgsFeatureId incrementalFeatureId )
```

File: src/core/providers/ogr/qgsogrprovider.cpp [hunk 2]
```diff
@@ -1860,16 +1860,23 @@ bool QgsOgrProvider::addFeaturePrivate( QgsFeature &f, Flags flags )
...
+      {
+        f.setId( id );
 
-      if ( mFirstFieldIsFid && attributes.count() > 0 )
+        if ( mFirstFieldIsFid && attributes.count() > 0 )
+        {
+          f.setAttribute( 0, id );
+        }
+      }
+      else if ( featureId >= 0 )
```
Comment (rouault) [inline src/core/providers/ogr/qgsogrprovider.cpp:1877]:
```suggestion
      else if ( incrementalFeatureId >= 0 )
```

File: src/core/providers/ogr/qgsogrprovider.cpp [hunk 3]
```diff
@@ -1860,16 +1860,23 @@ bool QgsOgrProvider::addFeaturePrivate( QgsFeature &f, Flags flags )
...
       {
-        f.setAttribute( 0, id );
+        f.setId( featureId );
```
Comment (rouault) [inline src/core/providers/ogr/qgsogrprovider.cpp:1879]:
```suggestion
        f.setId( incrementalFeatureId );
```

File: src/core/providers/ogr/qgsogrprovider.cpp [hunk 4]
```diff
@@ -1890,13 +1897,32 @@ bool QgsOgrProvider::addFeatures( QgsFeatureList &flist, Flags flags )
 
   const bool inTransaction = startTransaction();
 
+  QgsFeatureId incrementalFeatureId = -1;
+  OGRGeometryH filter;
```
Comment (rouault) [inline src/core/providers/ogr/qgsogrprovider.cpp:1901]: below block could be restricted to ``if ( !( flags & QgsFeatureSink::FastInsert ) )``

File: src/core/providers/ogr/qgsogrprovider.h [hunk 5]
```diff
@@ -313,7 +313,7 @@ class QgsOgrProvider final: public QgsVectorDataProvider
     //! converts \a value from json QVariant to QString
     QString jsonStringValue( const QVariant &value ) const;
 
-    bool addFeaturePrivate( QgsFeature &f, QgsFeatureSink::Flags flags );
+    bool addFeaturePrivate( QgsFeature &f, QgsFeatureSink::Flags flags, QgsFeatureId featureId = 0 );
```
Comment (rouault) [inline src/core/providers/ogr/qgsogrprovider.h:316]:
```suggestion
    bool addFeaturePrivate( QgsFeature &f, QgsFeatureSink::Flags flags, QgsFeatureId incrementalFeatureId = -1 );
```

**Comment by nirvn** [comment]
@rouault , thanks for the review, all comments addressed.

**Comment by nirvn** [comment]
@rouault , OK for the adjusted commit to be merged?

**Pull request merged**